src/boot/SerializeEscaped.js

```diff
@@ -0,0 +1,155 @@
+ * in its built-in replacers and revivers.
+ */
+export const SERIALIZED_TYPE_FIELD_NAME: '__serializedType__' = '__serializedType__';
+
+/**
+ * The single character to use in the escape strategy, as a prefix.
+ *
+ * This is not meant to be changed. We declare this `const` variable
+ * to clarify the escape/unescape patterns and enforce consistency
+ * between them.
+ *
+ * If you need to change this at all, be sure to write a migration. If
+ * you need to make it more than one character, also rename the
+ * variable and modify the escape/unescape patterns accordingly.
+ *
+ * Exported only for tests.
+ */
+export const ESCAPE_CHARACTER: '!' = '!';
+
+type StringTransform = (oldString: string) => string;
+
+/**
+ * Escape a string matching /^(!*)__serializedType__$/ as /^!\1__serializedType__$/
+ *
+ * Exported only for tests.
+ */
+export const escape: StringTransform = (key: string) =>
+  key.replace(
+    new RegExp('^(' + ESCAPE_CHARACTER + '*)' + SERIALIZED_TYPE_FIELD_NAME + '$'),
+    (match, p1) => ESCAPE_CHARACTER + p1 + SERIALIZED_TYPE_FIELD_NAME,
+  );
+
+/**
+ * Inverse of `escape`.
+ *
+ * Exported only for tests.
+ */
+export const unescape: StringTransform = (key: string) =>
+  key.replace(
+    new RegExp(
+      '^' + ESCAPE_CHARACTER + '(' + ESCAPE_CHARACTER + '*)' + SERIALIZED_TYPE_FIELD_NAME + '$',
+    ),
+    (match, p1) => p1 + SERIALIZED_TYPE_FIELD_NAME,
+  );
+
+type InexactObject = { [key: string]: mixed };
+
+type ObjectKeyRenamer = (obj: InexactObject) => InexactObject;
+
+const makeObjectKeyRenamer = (renameFn: StringTransform): ObjectKeyRenamer =>
+  /**
+   * Rename all keys of an object with the supplied renameFn.
+   *
+   * Does not mutate the object. If no names actually change, returns
+   * the same object; otherwise, makes a new object and returns that.
+   * Original ordering of keys is not preserved in this new object.
+   */
+  (obj: InexactObject): InexactObject => {
+    const renamedEntries: $ElementType<$Call<typeof objectEntries, InexactObject>, number>[] = [];
+    const untouchedEntries: $ElementType<$Call<typeof objectEntries, InexactObject>, number>[] = [];
+    for (const entry of objectEntries(obj)) {
+      const [key, _] = entry; /* eslint-disable-line no-unused-vars */
+      const newName = renameFn(key);
+      if (key !== newName) {
+        entry[0] = newName; // *elements* of a $ReadOnlyArray may be mutable
+        renamedEntries.push(entry);
+      } else {
+        untouchedEntries.push(entry);
+      }
+    }
+    if (renamedEntries.length > 0) {
+      return {
+        // Object.fromEntries is added in facebook/flow@f92231cf1,
+        // released in v0.96.0. Until we use that, $FlowFixMe.
+        ...Object.fromEntries(untouchedEntries),
+        // $FlowFixMe
+        ...Object.fromEntries(renamedEntries),
+      };
+    } else {
+      return obj;
+    }
+  };
+
+const escapeKeys = makeObjectKeyRenamer(escape);
```

Comment:
I wouldn't insist on using a `makeObjectKeyRenamer` factory; we could just have a `renameObjectKeys(obj, renameFn)` and pass `escape` or `unescape` as `renameFn` every time. But, since `escapeKeys` and `unescapeKeys` are used several times, I think it's more readable that they exist as separate functions, with those names, and a factory function was a natural way to accomplish that.

Reply:
> But, since `escapeKeys` and `unescapeKeys` are used several times

— er, no they're not! 🙃 I must have been thinking of a previous iteration I never submitted for review.